openrouteservice/src/main/java/heigit/ors/routing/graphhopper/extensions/core/CoreDijkstra.java

```diff
@@ -0,0 +1,270 @@
+            if (inCore)
+                visitedCountFrom2++;
+            else
+                visitedCountFrom1++;
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean fillEdgesTo() {
+        if (pqCHTo.isEmpty())
+            return false;
+
+        currTo = pqCHTo.poll();
+
+        if (!inCore && chGraph.getLevel(currTo.adjNode) == coreNodeLevel) {
+            // core entry point, do not relax its edges
+            pqCoreTo.add(currTo);
+        }
+        else {
+            bestWeightMapOther = bestWeightMapFrom;
+            fillEdges(currTo, pqCHTo, bestWeightMapTo, inEdgeExplorer, true);
+            if (inCore)
+                visitedCountTo2++;
+            else
+                visitedCountTo1++;
+        }
+
+        return true;
+    }
+
+    @Override
+    public boolean finishedPhase1() {
+        // we need to finish BOTH searches for CH!
+        if (finishedFrom && finishedTo)
+            return true;
+
+        double fromWeight = currFrom.weight;
+        double toWeight = currTo.weight;
+
+        // changed also the final finish condition for CH
+        if (!pqCoreFrom.isEmpty())
+            fromWeight = Math.min(pqCoreFrom.peek().weight, fromWeight);
+        if (!pqCoreTo.isEmpty())
+            toWeight = Math.min(pqCoreTo.peek().weight, toWeight);
+
+
+        return fromWeight >= bestPath.getWeight() && toWeight >= bestPath.getWeight();
+    }
+
+    @Override
+    void runPhase2() {
+        // re-init queues
+        pqCHFrom = pqCoreFrom;
+        pqCHTo = pqCoreTo;
+
+        finishedFrom = pqCHFrom.isEmpty();
+        finishedTo = pqCHFrom.isEmpty();
+
+        if (!finishedFrom)
+            currFrom = pqCHFrom.peek();
+
+        if (!finishedTo)
+            currTo = pqCHTo.peek();
+
+        while (!finishedPhase2() && !isMaxVisitedNodesExceeded()) {
+            if (!finishedFrom)
+                finishedFrom = !fillEdgesFrom();
+
+            if (!finishedTo)
+                finishedTo = !fillEdgesTo();
+        }
+    }
+
+    @Override
+    public boolean finishedPhase2() {
+        if (finishedFrom || finishedTo)
+            return true;
+
+        return currFrom.weight + currTo.weight >= bestPath.getWeight();
+    }
+
+    void fillEdges(SPTEntry currEdge, PriorityQueue<SPTEntry> prioQueue, IntObjectMap<SPTEntry> bestWeightMap,
+                   EdgeExplorer explorer, boolean reverse) {
+        EdgeIterator iter = explorer.setBaseNode(currEdge.adjNode);
+        while (iter.next()) {
+            if (!accept(iter, currEdge.edge))
+                continue;
+
+            int traversalId = traversalMode.createTraversalId(iter, reverse);
+            // Modification by Maxim Rylov: use originalEdge as the previousEdgeId
+            double tmpWeight = weighting.calcWeight(iter, reverse, currEdge.originalEdge) + currEdge.weight;
+            if (Double.isInfinite(tmpWeight))
+                continue;
+            SPTEntry ee = bestWeightMap.get(traversalId);
```

Comment:
This method uses a modified code from original GH implementation. I prefer to keep the original variable names to be able to quickly cross-compare so see the changes we have introduced. I believe this will help in case we decide to move the implementation over to GH.